### ISSUE TITLE
Question regarding anchors.

### ISSUE THREAD

**Issue by Moondra** [experiencor/basic-yolo-keras#135]
I have a general question regarding anchors and how they are implemented in YOLO.
It seems each anchor only needs two numbers? If so how is an anchor box created with only two numbers? Our ground truth labels require 4 numbers, so I'm confused as to how an anchor box can be
created with only two numbers. 

My second question:
In your code, it seems you generate anchor boxes automatically.
I don't quite understand what varaiables you are using in the IOU function.
As for your IOU function, are the `centroids` the ground truth centroids?
Is `ann`anchor? 

```
def IOU(ann, centroids):
    w, h = ann
    similarities = []

    for centroid in centroids:
        c_w, c_h = centroid

```

Thank you.

**Comment by experiencor** [comment]
@Moondra We only need the width and the height of the anchor. So we do clustering on the widths and heights of the objects.

```centroids``` is the list of the current best anchors' sizes (w and h) and ```ann``` is the size of the object in investigation. This function computes the IOU distances between that object and all the anchors.

**Comment by moondra2017** [comment]
@experiencor Thank you. If we only have the width and height of the anchor box, what is the starting point of the anchor box? Our height and width have to be based upon some center point right? Is it based on our ground truth's center points (bx, by?).  

So from what I'm assuming, we get a list of anchor points (width and height). We then use those to create boxes based off some point( ground truth center point). We then calculate the IOU between ground truth box and anchor boxes over all examples. Average the IOU and pick the best anchor points (height and width). Is this how it works? 

Forgive me for all the questions, I'm going through your code as well as the original paper, and am having a hard time understanding all the details. Thank you so much.

**Comment by experiencor** [comment]
@moondra2017 The starting point of the anchor is identical to the starting point of the object. Each makes 5 prediction tuple (x,y,w,h,confidence,class). The 5 predicted (w,h)'s are then scaled using the anchors' sizes. The 5 predicted (x,y)'s are left untouched. You can think of them as the 5 positions of the 5 anchors.

I'm not sure if you are asking the training process or the clustering process.

During the training, we need to find the best anchor for each bounding box. To do this, we use the position of the bounding box as the position of the 5 anchors. We then compute the IOU of the bounding box and all 5 anchors. The anchor with the highest IOU is selected.

During the clustering of the bounding box sizes to find the best anchors, we simply do KMean on all pairs of (bounding box width,bounding box height) appeared in the dataset. IOU is the distance measure that we use (instead of Euclidean distance).

**Comment by Moondra** [comment]
＠experiencor Ah, I think I understand!  So the anchor coordinates are just making adjustments to the 
the prediction box (the 5 element tuple).

Just for clarity sake I've created an example. 
I'm mainly talking about the training phase, but let's say we have one anchor box( for simplicity sake).
Before starting training, we predetermine anchor box values [.09, .10] (using K-means)

After training on one example, say we get back the prediction as (.30,.30,.6,.6,1,1)

So to create the anchor box above, we adjust the w, h by .09 and .10 (subtract)?
Creating something like :   (.30,.30, .6 - .09, .6 - .10)
which gives us   (.30, .30, .51, ,50)

With these four coordinates we can draw an anchor box. 
We then easily check IOU of this anchor box with the ground truth box.

Thank you so much for the explanation (I went through Andrew Ng's coursea, course which introduced YOLO, but a lot of details were left out regarding anchor boxes).
